extension/healthcheckv2extension/internal/http/server_test.go

```diff
@@ -0,0 +1,4089 @@
+							metrics.ExporterID,
+							component.NewPermanentErrorEvent(assert.AnError),
+						)
+					},
+					queryParams:        "verbose",
+					expectedStatusCode: http.StatusInternalServerError,
+					expectedComponentStatus: &componentStatusExpectation{
+						healthy: false,
+						status:  component.StatusPermanentError,
+						err:     assert.AnError,
+						nestedStatus: map[string]*componentStatusExpectation{
+							"pipeline:traces": {
+								healthy: true,
+								status:  component.StatusOK,
+								nestedStatus: map[string]*componentStatusExpectation{
+									"receiver:traces/in": {
+										healthy: true,
+										status:  component.StatusOK,
+									},
+									"processor:batch": {
+										healthy: true,
+										status:  component.StatusOK,
+									},
+									"exporter:traces/out": {
+										healthy: true,
+										status:  component.StatusOK,
+									},
+								},
+							},
+							"pipeline:metrics": {
+								healthy: false,
+								status:  component.StatusPermanentError,
+								nestedStatus: map[string]*componentStatusExpectation{
+									"receiver:metrics/in": {
+										healthy: true,
+										status:  component.StatusOK,
+									},
+									"processor:batch": {
+										healthy: true,
+										status:  component.StatusOK,
+									},
+									"exporter:metrics/out": {
+										healthy: false,
+										status:  component.StatusPermanentError,
+										err:     assert.AnError,
+									},
+								},
+							},
+						},
+					},
+				},
+				{
+					queryParams:        "pipeline=metrics&verbose",
+					expectedStatusCode: http.StatusInternalServerError,
+					expectedComponentStatus: &componentStatusExpectation{
+						healthy: false,
+						status:  component.StatusPermanentError,
+						err:     assert.AnError,
+						nestedStatus: map[string]*componentStatusExpectation{
+							"receiver:metrics/in": {
+								healthy: true,
+								status:  component.StatusOK,
+							},
+							"processor:batch": {
+								healthy: true,
+								status:  component.StatusOK,
+							},
+							"exporter:metrics/out": {
+								healthy: false,
+								status:  component.StatusPermanentError,
+							},
+						},
+					},
+				},
+				{
+					step: func() {
+						testhelpers.SeedAggregator(
+							server.aggregator,
+							traces.InstanceIDs(),
+							component.StatusStopping,
+						)
+						testhelpers.SeedAggregator(
+							server.aggregator,
+							metrics.InstanceIDs(),
+							component.StatusStopping,
+						)
+					},
+					queryParams:        "verbose",
+					expectedStatusCode: http.StatusServiceUnavailable,
+					expectedComponentStatus: &componentStatusExpectation{
+						healthy: true,
+						status:  component.StatusStopping,
+						nestedStatus: map[string]*componentStatusExpectation{
+							"pipeline:traces": {
+								healthy: true,
+								status:  component.StatusStopping,
+								nestedStatus: map[string]*componentStatusExpectation{
```

Comment:
might be worth adding a reusable variable for the common nestedStatus values for each pipelines

Reply:
here's the suggestion for refactoring https://github.com/mwear/opentelemetry-collector-contrib/pull/12294

Reply:
thanks for the suggestion. i merged it and found other places the strategy could be applied.